carsus/io/output/base.py

```diff
@@ -0,0 +1,546 @@
+        levels_w_ionization_energies = pd.merge(levels_all,
+                                                ionization_energies,
+                                                how='left',
+                                                on=["atomic_number",
+                                                    "ion_number"])
+        mask = levels_w_ionization_energies["energy"] < \
+            levels_w_ionization_energies["ionization_energy"]
+        levels = levels_w_ionization_energies[mask].copy()
+        levels = levels.set_index('level_id').sort_values(
+            by=['atomic_number', 'ion_number'])
+        levels = levels.drop(columns='ionization_energy')
+
+        # Clean lines
+        lines = lines_all.join(pd.DataFrame(index=levels.index),
+                               on="lower_level_id", how="inner").\
+            join(pd.DataFrame(index=levels.index),
+                 on="upper_level_id", how="inner")
+
+        # Culling lines with low gf values
+        lines = lines.loc[lines["loggf"] > lines_loggf_threshold]
+
+        # Do not clean levels that don't exist in lines
+
+        # Create the metastable flags for levels
+        levels["metastable"] = \
+            self._create_metastable_flags(levels, lines_all,
+                                          levels_metastable_loggf_threshold)
+
+        # Create levels numbers
+        levels = levels.sort_values(
+            ["atomic_number", "ion_number", "energy", "g"])
+        levels["level_number"] = levels.groupby(['atomic_number',
+                                                 'ion_number'])['energy']. \
+            transform(lambda x: np.arange(len(x))).values
+        levels["level_number"] = levels["level_number"].astype(np.int)
+
+        levels = levels[['atomic_number', 'energy', 'g', 'ion_number',
+                         'level_number', 'metastable']]
+
+        # Join atomic_number, ion_number, level_number_lower,
+        # level_number_upper on lines
+        lower_levels = levels.rename(
+            columns={
+                "level_number": "level_number_lower",
+                "g": "g_l"}
+        ).loc[:, ["atomic_number", "ion_number", "level_number_lower", "g_l"]]
+        upper_levels = levels.rename(
+            columns={
+                "level_number": "level_number_upper",
+                "g": "g_u"}
+        ).loc[:, ["level_number_upper", "g_u"]]
+        lines = lines.join(lower_levels, on="lower_level_id").join(
+            upper_levels, on="upper_level_id")
+
+        # Calculate absorption oscillator strength f_lu and emission
+        # oscillator strength f_ul
+        lines["f_lu"] = lines["gf"] / lines["g_l"]
+        lines["f_ul"] = lines["gf"] / lines["g_u"]
+
+        # Calculate frequency
+        lines['nu'] = u.Quantity(
+            lines['wavelength'], 'angstrom').to('Hz', u.spectral())
+
+        # Calculate Einstein coefficients
+        einstein_coeff = (4 * np.pi ** 2 * const.e.gauss.value **
```

Comment:
We may want to move the calculation of the Einstein coefficients into a staticmethod. This method is already very  long.

Reply:
Added to TODO list

Reply:
Done